### PR TITLE
add input type and dtype check for accuracy_op 

### PR DESCRIPTION
**Pull request by JesseyXujin**

### CODE REVIEW

**Comment by JesseyXujin** [comment]
**fluid.layers.accuracy**
+ Check accuracy shape
```python
import paddle.fluid as fluid
import numpy as np
x1 = fluid.layers.data(name='input', shape=[-1,1], dtype="float32")
label = fluid.layers.data(name='label', shape=[-1], dtype="int32")
fluid.layers.accuracy(x1,label=label)
```
**before:** `label's rank must be 2`
**after:** `ShapeError: label's dimensions of AccuracyOp must be 2. But received label's dimensions = 1`
```python
import paddle.fluid as fluid
import numpy as np
x1 = fluid.layers.data(name='input', shape=[-1,1], dtype="float32")
label = fluid.layers.data(name='label', shape=[-1,2], dtype="int32")
fluid.layers.accuracy(x1,label=label)
```
**before:** `label's second dimension must be 1`
**after:** `ShapeError: label's second dimension of AccuracyOp must be 1. But received label's second dimension is = 2.`
```python
import paddle.fluid as fluid
import numpy as np
data = fluid.layers.data(name="input", shape=[-1, 32, 32], dtype="float16")
label = fluid.layers.data(name="label", shape=[-1,1], dtype="int32")
fc_out = fluid.layers.fc(input=data, size=10)
predict = fluid.layers.softmax(input=fc_out)
result = fluid.layers.accuracy(input=predict, label=label, k=5)
place = fluid.CPUPlace()
exe = fluid.Executor(place)
exe.run(fluid.default_startup_program())
x = np.random.rand(3, 32, 32).astype("float16")
y = np.array([[1],[0],[1]])
output= exe.run(feed={"input": x,"label": y},
                 fetch_list=[result[0]])
print(output)
```
**before:** `the inference tensor's num_rows must be the same as label.`
**after:** `ShapeError:the output's num_rows of AccuracyOp must be the same as label's num_rows. But received output's shape = [1, 32, 1], label's shape = [3, 1], output's num_rows = 1, label's num_rows = 3`

File: paddle/fluid/operators/metrics/accuracy_op.cc [hunk 1]
```diff
@@ -40,12 +40,22 @@ class AccuracyOp : public framework::OperatorWithKernel {
...
+    PADDLE_ENFORCE_EQ(label_dim.size(), 2,
+                      "ShapeError: label's dimensions of AccuracyOp must be 2. "
+                      "But received label's dimensions = %d",
+                      label_dim.size());
     PADDLE_INFERSHAPE_ENFORCE_EQ(ctx, label_dim[1], 1,
-                                 "label's second dimension must be 1");
-    PADDLE_INFERSHAPE_ENFORCE_EQ(ctx, inference_dim[0], label_dim[0],
-                                 "the inference tensor's num_rows must be"
-                                 " the same as label.");
+                                 "ShapeError: label's second dimension of "
```
Comment (xyzhou-puck) [inline paddle/fluid/operators/metrics/accuracy_op.cc]: embedding，one-hot等op已经去掉了这个要求，可以确认一下是否ok。

**Review by xyzhou-puck — DISMISSED** [review]
LGTM

File: python/paddle/fluid/layers/metric_op.py [hunk 2]
```diff
@@ -71,6 +72,14 @@ def accuracy(input, label, k=1, correct=None, total=None):
             #[array([0.6666667], dtype=float32)]
     """
     helper = LayerHelper("accuracy", **locals())
+    if not isinstance(input, Variable):
+        raise TypeError(
+            "The type of 'input' in softmax must be Variable, but received %s" %
```
Comment (chenwhql) [inline python/paddle/fluid/layers/metric_op.py]: softmax -> accuracy

File: paddle/fluid/operators/metrics/accuracy_op.cc [hunk 3]
```diff
@@ -40,12 +40,22 @@ class AccuracyOp : public framework::OperatorWithKernel {
     // Assume indices has same shape as inference, because
     // it's the output of topk.
 
-    PADDLE_ENFORCE_EQ(label_dim.size(), 2, "label's rank must be 2.");
+    PADDLE_ENFORCE_EQ(label_dim.size(), 2,
+                      "ShapeError: label's dimensions of AccuracyOp must be 2. "
```
Comment (chenwhql) [inline paddle/fluid/operators/metrics/accuracy_op.cc]: 请遵守规则，打印出的具体的dim

image

参照示例：https://github.com/PaddlePaddle/Paddle/pull/19998

File: paddle/fluid/operators/metrics/accuracy_op.cc [hunk 4]
```diff
@@ -40,12 +40,22 @@ class AccuracyOp : public framework::OperatorWithKernel {
...
+                      label_dim.size());
     PADDLE_INFERSHAPE_ENFORCE_EQ(ctx, label_dim[1], 1,
-                                 "label's second dimension must be 1");
-    PADDLE_INFERSHAPE_ENFORCE_EQ(ctx, inference_dim[0], label_dim[0],
-                                 "the inference tensor's num_rows must be"
-                                 " the same as label.");
+                                 "ShapeError: label's second dimension of "
+                                 "AccuracyOp must be 1. But received label's "
+                                 "second dimension is = %d",
+                                 label_dim[1]);
```
Comment (chenwhql) [inline paddle/fluid/operators/metrics/accuracy_op.cc]: 同上，请打印`label_dim`

File: paddle/fluid/operators/metrics/accuracy_op.cc [hunk 5]
```diff
@@ -40,12 +40,22 @@ class AccuracyOp : public framework::OperatorWithKernel {
...
+    PADDLE_INFERSHAPE_ENFORCE_EQ(
+        ctx, inference_dim[0], label_dim[0],
+        "ShapeError:the output's num_rows of AccuracyOp must be"
```
Comment (chenwhql) [inline paddle/fluid/operators/metrics/accuracy_op.cc]: `ShapeError:`后面增加一个空格

**Review by chenwhql — COMMENTED** [review]
accuracy_op.h，这个文件没有实际修改，为什么要提交？建议checkout回去

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by chenwhql — CHANGES_REQUESTED** [review]
accuracy op是支持float1.6的，请增加python端类型的检查

File: python/paddle/fluid/layers/metric_op.py [hunk 6]
```diff
@@ -71,6 +72,14 @@ def accuracy(input, label, k=1, correct=None, total=None):
             #[array([0.6666667], dtype=float32)]
     """
     helper = LayerHelper("accuracy", **locals())
+    if not isinstance(input, Variable):
+        raise TypeError(
+            "The type of 'input' in accuracy must be Variable, but received %s"
+            % (type(input)))
+    if convert_dtype(input.dtype) not in ['float16', 'float32', 'float64']:
```
Comment (chenwhql) [inline python/paddle/fluid/layers/metric_op.py:83]: 请添加fp16的warning提示

**Review by chenwhql — APPROVED** [review]

**Pull request merged**